tests/linearizability/model.go

```diff
@@ -187,24 +198,32 @@ func stepDelete(state EtcdState, request EtcdRequest, response EtcdResponse) (bo
 		state.FailedWrite = &request
 		return true, state
 	}
-	// revision should never decrease
-	if response.Revision < state.LastRevision {
-		return false, state
-	}
 	deleteSucceeded := response.Deleted != 0
 	keySet := state.Value != ""
 
-	// non-existent key cannot be deleted.
-	if deleteSucceeded != keySet && state.FailedWrite == nil {
-		return false, state
-	}
-	//if key was deleted, response revision should increase
-	if deleteSucceeded && (response.Revision != state.LastRevision+1 || !keySet) && (state.FailedWrite == nil || response.Revision < state.LastRevision+2) {
-		return false, state
-	}
-	//if key was not deleted, response revision should not change
-	if !deleteSucceeded && state.LastRevision != response.Revision && state.FailedWrite == nil {
-		return false, state
+	if state.FailedWrite == nil {
+		// non-existent key cannot be deleted.
+		if deleteSucceeded != keySet {
+			return false, state
+		}
+		// if key was deleted, response revision should increase by one
+		if deleteSucceeded && response.Revision != state.LastRevision+1 {
+			return false, state
+		}
+		// if key was not deleted, response revision should not change
+		if !deleteSucceeded && state.LastRevision != response.Revision {
+			return false, state
+		}
+	} else {
+		// if key was set, revision must increase as either key was deleted or failed request was persisted.
+		if keySet && response.Revision < state.LastRevision+1 {
+			return false, state
+		}
+
+		// if delete succeeded (+1 rev) without any key, means that fail request was persisted (+1 rev).
+		if deleteSucceeded && !keySet && response.Revision < state.LastRevision+2 {
+			return false, state
+		}
 	}
 
 	state.Value = ""
```

Comment:
We need to cleanup `FailedWrite` on successful delete operation.
```
state.FailedWrite = nil
```

Reply:
You are right.

Reply:
I think that `state.FailedWrite = nil` should be set even if delete failed. No matter if delete failed or not, we know final state. Value will be empty no matter what.

Reply:
> we know final state. Value will be empty no matter what

Do not understand this. If the client gets an error response, how can you know the real server side state?